### PR TITLE
[Feat] Added VideoSequential for video augmentations

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
#802

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/augmentation/container/video.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        # Convert to (B, T, C, H, W)
+        input = input.transpose(1, 2)
+        _original_shape = input.shape
+        input = input.reshape(-1, *batch_shape[1:])
+        if not self.same_on_frame:
+            # Overwrite param generation shape to (B * T, C, H, W).
+            batch_shape = input.shape
+        # for aug in self.augmentation_list:
+        for aug in self.children():
+            param = aug.generate_parameters(batch_shape)
```
Comment (jbohnslav) [inline kornia/augmentation/container/video.py:90]: Note: I didn't want to fork your fork, so here's a quick fix: 
```python
if 'batch_prob' not in param.keys():
    # print(param)
    probability = aug.__forward_parameters__(batch_shape, aug.p, aug.p_batch, aug.same_on_batch)
    param.update(probability)
```

**Review by jbohnslav — CHANGES_REQUESTED** [review]
There's a bug where the batch probability is not actually used. Because we're using `aug(..., params=params)`, the forward function doesn't generate batch probabilities implicity. This means that when I used `colorjitter`, `RandomAffine`, `RandomGrayscale`, and potentially others, it was applying the transform to every image. 

Here's a batch with RandomGrayscale(p=0.01) and RandomAffine(p=0.01) before the fix. 
image

File: kornia/augmentation/container/video.py (hunk 1 above)
Comment (shijianjian) [reply kornia/augmentation/container/video.py]: Good catch. It has been fixed.

**Review by jbohnslav — APPROVED** [review]
Works for me!

**Comment by shijianjian** [comment]
> Works for me! Thanks for your work on this, it is a great solution. 

Brilliant. This is PR is going to be further evaluated and discussed.

We wander if to support more things like both BCTHW and BTCHW. Please let us know if you got any further ideas and use cases.

**Comment by edgarriba** [comment]
Indeed, we had a quick discussion with @lferraz pointing the following things:
- The data format for video should either `(*, C, H, W)` or `(*, H, W, C)` 
  - the last because of torchvision returning `HWC` video frames
However, to follow the same convention across kornia we should expect by default `(*,C,H,W)`.
The proposal in this PR fits the use case for using conv3d, but I wouldn't say that it will be the most common use case for applying transforms to video sequences.

@shijianjian apart from what I just mentioned, your current proposal in this PR could be slightly inefficient since we intrinsically perform a permutation in the channels to apply the transforms and then again permute to return for `BCTHW`. Besides, in an end-to-end application there would be the extra cost of permuting from `HWC` from torchvision (or I believe cv2).

To summarise, I would force this API to expect `(*, C, H, W)` as input tensors to avoid confusions with the rest of the library. Any other opinions ? @jbohnslav /cc  @ducha-aiki @anguelos

**Comment by jbohnslav** [comment]
Indeed, cv2 and torchvision return `HWC`, but PyTorch expects `TCHW` in the `Conv3D`. For example, in my CPU-based dataloaders, I always return `BCTHW`. If we switched over to force this module to expect `*CHW`, then I would always have to use: `nn.Sequential(TransposeToBTCHW, VideoSequential, TransposeToBCTHW)`. I prefer the current version.

File: kornia/augmentation/container/video.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        batch_shape = self.__infer_channel_exclusive_batch_shape__(input)
+        # Convert to (B, T, C, H, W)
+        input = input.transpose(1, 2)
+        _original_shape = input.shape
+        input = input.reshape(-1, *batch_shape[1:])
+        if not self.same_on_frame:
+            # Overwrite param generation shape to (B * T, C, H, W).
+            batch_shape = input.shape
+        for aug in self.children():
+            aug = cast(_AugmentationBase, aug)
```
Comment (jbohnslav) [inline kornia/augmentation/container/video.py:90]: I tried to put a Normalize into this class, and it threw an error because it doesn't have a __forward_parameters__. 

What do you think about: 
```python
if isinstance(aug, _AugmentationBase):  
  param = aug.__forward_parameters__(batch_shape, aug.p, aug.p_batch, aug.same_on_batch)
  if self.same_on_frame:
    for k, v in param.items():
      # TODO: revise colorjitter order param in the future to align the standard.
      if not (k == "order" and isinstance(aug, ColorJitter)):
        param.update({k: self.__repeat_param_across_channels__(v, frame_num)})
  input = aug(input, params=param)
else:
  assert isinstance(aug, nn.Module)
  input = aug(input)
```

Comment (shijianjian) [reply kornia/augmentation/container/video.py:90]: I prefer rewriting the Normalize/Denornalize to align the current APIs.

From I was planning, we may have more advanced features in this API, like passing param directly to it, etc. Still, we need to evaluate the most common use cases first.

**Comment by shijianjian** [comment]
> Indeed, cv2 and torchvision return `HWC`, but PyTorch expects `TCHW` in the `Conv3D`. For example, in my CPU-based dataloaders, I always return `BCTHW`. If we switched over to force this module to expect `*CHW`, then I would always have to use: `nn.Sequential(TransposeToBTCHW, VideoSequential, TransposeToBCTHW)`. I prefer the current version. 

Right. Currently, the only purpose of this VideoSequential is to do the automatic transposes internally. Thus, I would say to have something like "VideoSequential(..., format='BCTHW')" to support different use cases.

File: kornia/augmentation/functional/functional.py [hunk 3]
```diff
@@ -983,3 +983,11 @@ def apply_cutmix(input: torch.Tensor, labels: torch.Tensor,
             labels.to(input.dtype), labels_permute.to(input.dtype), lam.to(labels.device)], dim=1))
 
     return out_inputs, torch.stack(out_labels, dim=0)
+
+
+def apply_normalize(input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
+    return (input - params['mean']) / params['std']
```
Comment (edgarriba) [inline kornia/augmentation/functional/functional.py:989]: why not directly call the function kornia.enhance.normalize in the main class ?

File: kornia/enhance/normalize.py [hunk 4]
```diff
@@ -127,6 +129,73 @@ def normalize(
     return out
 
 
+def normalize3d(
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:132]: add tests for this

File: kornia/enhance/normalize.py [hunk 5]
```diff
@@ -239,6 +308,73 @@ def denormalize(
     return out
 
 
+def denormalize3d(
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:311]: add tests for this

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: test/augmentation/test_container.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        K.RandomMotionBlur(3, 35., 0.5, p=1.),
+        K.Normalize(torch.tensor([0.5, 0.5, 0.5]), torch.tensor([0.5, 0.5, 0.5]), p=1.),
+        K.Denormalize(torch.tensor([0.5, 0.5, 0.5]), torch.tensor([0.5, 0.5, 0.5]), p=1.),
+    ])
+    @pytest.mark.parametrize('data_format', ["BCTHW", "BTCHW"])
+    def test_augmentation(self, augmentation, data_format, device, dtype):
+        input = torch.randint(255, (1, 3, 3, 5, 6), device=device, dtype=dtype).repeat(2, 1, 1, 1, 1) / 255.
+        torch.manual_seed(21)
+        aug_list = K.VideoSequential(augmentation, data_format=data_format, same_on_frame=True)
+        output = aug_list(input)
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply test/augmentation/test_container.py:43]: To test if all augmentations can be safely forward. Debugged bounding box issues using this.

File: kornia/augmentation/functional/functional.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/functional/functional.py:989]: That is only for 2D, but I use this for both 2D and 3D. Also, the shape check of ```kornia.enhance.normalize``` is a bit buggy. I will refine that later when I got some more time.

[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/functional/functional.py:989]: BTW, both normalize and denormalize shall be shape-agnostic which means to be working with any shaped data as long as the channel dim is 1.

File: test/augmentation/test_augmentation_3d.py [hunk 7]
```diff
@@ -18,980 +18,1098 @@
...
     CenterCrop3D,
-    RandomEqualize3D
+    RandomEqualize3D,
+    Normalize3D,
+    Denormalize3D,
 )
 
 
-class TestRandomHorizontalFlip3D:
+# class TestRandomHorizontalFlip3D:
```
Comment (edgarriba) [inline test/augmentation/test_augmentation_3d.py:27]: why all this code is commented ? we shouldn't include commented code

Comment (shijianjian) [reply test/augmentation/test_augmentation_3d.py]: Sorry, forgot to uncomment it out.

File: test/augmentation/test_augmentation_3d.py [hunk 8]
```diff
@@ -24,979 +24,979 @@
 )
 
 
-# class TestRandomHorizontalFlip3D:
```
Comment (edgarriba) [inline test/augmentation/test_augmentation_3d.py]: @shijianjian but what's the problem with this tests ?

Comment (shijianjian) [reply test/augmentation/test_augmentation_3d.py]: No problems on those. Only a typo actually. I commented other tests to play with the new tests.

**Comment by shijianjian** [comment]
I cannot reproduce the CUDA errors with 10.1, torch 1.7 and python 3.7. I suspect they are CUDA 10.2 specific.

**Comment by edgarriba** [comment]
@shijianjian cuda errors it's fine I think. The GPU server is quite busy these days and seems that might be the issue

File: kornia/enhance/normalize.py [hunk 9]
```diff
@@ -71,12 +71,12 @@ def normalize(
...
 
     Args:
-        data (torch.Tensor): Image tensor of size :math:`(*, C, H, W)`.
+        data (torch.Tensor): Image tensor of size :math:`(*, C, ...)`.
         mean (Union[torch.Tensor, float]): Mean for each channel.
         std (Union[torch.Tensor, float]): Standard deviations for each channel.
 
     Return:
-        torch.Tensor: Normalised tensor with same size as input :math:`(*, C, H, W)`.
+        torch.Tensor: Normalised tensor with same size as input :math:`(*, C, ...)`.
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:79]: @shijianjian do we have enough unit tests checking this ?

**Review by edgarriba — APPROVED** [review]
LGTM. Just verify with test that the normalize do we you expect

**Comment by edgarriba** [comment]
@shijianjian Please, add to docs and potentially short quick usage code.

**Comment by smivv** [comment]
Hello guys, sorry if I disturb you.
One quick question: is it possible now to use VideoSequential with bboxes/keypoints/masks?
I didn't find that. Thanks in advance!

**Comment by shijianjian** [comment]
Hi @smivv,

I haven't enabled it. But it should be easy to add by modifying:
https://github.com/kornia/kornia/blob/353fb68429b8b5eae4ad7c06389c0a1943935a3e/kornia/augmentation/container/augment.py#L361-L362

Then it can be used along with AugmentSequential:
```python
as = AugmentSequential(
    VideoSequential(...),
    data_keys=["input", "mask", "bbox", "keypoints"],
)
```

Can you play with it?

Jian

**Comment by smivv** [comment]
Hi @shijianjian,

Thanks for your reply! I tried to do the same as you proposed, but getting error even earlier here:

https://github.com/kornia/kornia/blob/353fb68429b8b5eae4ad7c06389c0a1943935a3e/kornia/augmentation/container/augment.py#L330

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Python/3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/Users/smivv/Library/Python/3.8/lib/python/site-packages/kornia/augmentation/container/augment.py", line 330, in forward
    _, out_shape = self.autofill_dim(inp, dim_range=(2, 4))
  File "/Users/smivv/Library/Python/3.8/lib/python/site-packages/kornia/augmentation/container/base.py", line 170, in autofill_dim
    raise RuntimeError(f"input shape expected to be in {dim_range} while got {ori_shape}.")
RuntimeError: input shape expected to be in (2, 4) while got torch.Size([8, 4, 3, 10, 10]).
```

So I changed top dim value to 5 and now getting another error like this:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Python/3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/Users/smivv/Library/Python/3.8/lib/python/site-packages/kornia/augmentation/container/augment.py", line 364, in forward
    raise NotImplementedError(f"data_key {dcate} is not implemented for {module}.")
NotImplementedError: data_key DataKey.KEYPOINTS is not implemented for VideoSequential(
  (RandomHorizontalFlip_0): RandomHorizontalFlip(p=0.5, p_batch=1.0, same_on_batch=False, return_transform=None)
).
```

So I believe there are plenty adjustments that has to be done :)

**Comment by shijianjian** [comment]
@smivv I will take a look shortly.

**Comment by smivv** [comment]
@shijianjian kindly asking to mention me in PR which will be about such functionality, thanks ❤️

**Comment by shijianjian** [comment]
> @shijianjian kindly asking to mention me in PR which will be about such functionality, thanks ❤️

Sorry for the delay. I got some personal things to deal with lately. I will start working on it in about one or two weeks.

If you need it in hurry and interested in working on it, you may join our slack and I can guide you there.
